File: cli/summary.py
from typing import Optional, List, Tuple
from rich.table import Table
from cli.console import console
from core.telegram_scanner import _fmt_speed as _tg_fmt, _fmt_size as _tg_size


def _format_summary(
    run_dns_avail: bool,
    run_domains: bool,
    run_tcp: bool,
    run_telegram: bool,
    domain_stats,
    tcp_stats,
    telegram_stats=None,
    dns_avail_stats=None,
) -> Optional[Table]:
    """Строит сводную таблицу результатов сканирования."""
    items: List[Tuple[str, str]] = []

    # ── Тест 1: доступность DNS ───────────────────────────────────────────────
    if run_dns_avail:
        if dns_avail_stats:
            d = dns_avail_stats
            doh_color = "green" if d["doh_ok"] == d["doh_total"] else (
                "red" if d["doh_ok"] == 0 else "yellow"
            )
            udp_color = "green" if d["udp_ok"] == d["udp_total"] else (
                "red" if d["udp_ok"] == 0 else "yellow"
            )
            dot_ok, dot_total = d.get("dot_ok", 0), d.get("dot_total", 0)
            dot_color = "green" if dot_ok == dot_total else (
                "red" if dot_ok == 0 else "yellow"
            )
            dns_parts = [f"[{doh_color}]{d['doh_ok']}/{d['doh_total']} DoH[/{doh_color}]"]
            if dot_total:
                dns_parts.append(f"[{dot_color}]{dot_ok}/{dot_total} DoT[/{dot_color}]")
            dns_parts.append(f"[{udp_color}]{d['udp_ok']}/{d['udp_total']} UDP[/{udp_color}]")
            items.append(("DNS доступность", "  ".join(dns_parts)))
            if d.get("hijacked_brands"):
                total_r = d.get("resolvers_total", 0)
                if total_r and len(d["hijacked_brands"]) >= total_r:
                    items.append(("Подмена резолвера", "[red]Все[/red]"))
                else:
                    items.append(("Подмена резолвера",
                        f"[red]{', '.join(d['hijacked_brands'])}[/red]"))
            else:
                items.append(("Подмена резолвера", "[dim]—[/dim]"))
            if d.get("subst_total"):
                fi = d.get("fakeip_sub", 0)
                if fi:
                    items.append(("FakeIP ответов",
                        f"у [magenta]{fi}/{d['subst_total']}[/magenta] UDP"))
                rest = d.get("subst_sub", 0) - fi
                if rest > 0:
                    color = "red" if rest == d["subst_total"] else "yellow"
                    items.append(("Подмена ответов",
                        f"у [{color}]{rest}/{d['subst_total']}[/{color}] UDP"))
                elif not fi:
                    items.append(("Подмена ответов",
                        f"у [green]0/{d['subst_total']}[/green] UDP"))
        else:
            items.append(("DNS доступность", "[dim]—[/dim]"))

    # ── Тест 2: домены ────────────────────────────────────────────────────────
    if domain_stats:
        d = domain_stats

        def _stat(label, ok):
            color = "green" if ok == d["total"] else ("red" if ok == 0 else "yellow")
            return f"[{color}]{ok}/{d['total']} {label}[/{color}]"

        items.append(("Домены",
            _stat("HTTP", d["http_ok"])
            + f"  {_stat('TLS1.2', d['t12_ok'])}"
            + f"  {_stat('TLS1.3', d['t13_ok'])}"))

    # ── Тест 3: TCP 16-20KB ───────────────────────────────────────────────────
    if tcp_stats:
        t = tcp_stats
        pct = int(t["ok"] / t["total"] * 100) if t["total"] else 0
        value = (
            f"[green]√ {t['ok']}/{t['total']} OK[/green]"
            + (f"  [red]× {t['blocked']} блок.[/red]" if t['blocked'] else "")
            + (f"  [yellow]≈ {t['mixed']} смеш.[/yellow]" if t['mixed'] else "")
            + f"  [dim]({pct}% ОК)[/dim]"
        )
        items.append(("TCP 16-20KB", value))

    # ── Тест 5: Telegram ──────────────────────────────────────────────────────
    if run_telegram and telegram_stats:
        t = telegram_stats
        dl_data = t.get("download", {})
        ul_data = t.get("upload", {})
        dc_r, dc_t = t.get("dc_reachable", 0), t.get("dc_total", 0)

        def _fmt_tg(label, data, speed_key, size_key):
            st   = data.get("status")
            avg  = data.get(speed_key, 0)
            size = data.get(size_key, 0)
            drop = data.get("drop_at_sec")
            if st == "ok":
                raw_st, color = "ОК", "green"
            elif st == "stalled":
                raw_st, color = "ЗАМЕДЛЕНИЕ+ОБРЫВ", "yellow"
            elif st == "slow":
                raw_st, color = "ЗАМЕДЛЕНИЕ", "yellow"
            elif st == "blocked":
                raw_st, color = "НЕДОСТУПНО", "red"
            else:
                raw_st, color = "ОШИБКА", "red"
            metrics = f"ср. {_tg_fmt(avg)}, {_tg_size(size)}"
            if drop:
                metrics += f", обрыв на {drop}с"
            return label, f"[{color}]{raw_st:<16}[/{color}] {metrics}"

        items.append(_fmt_tg("TG Скачивание", dl_data, "avg_bps", "bytes_total"))
        items.append(_fmt_tg("TG Загрузка",   ul_data, "bps",     "sent"))
        dc_color = "green" if dc_r == dc_t else ("red" if dc_r == 0 else "yellow")
        items.append(("TG Датацентры", f"[{dc_color}]ОК {dc_r}/{dc_t}[/{dc_color}]"))

    if not items:
        return None
    t = Table(show_header=False, box=None, pad_edge=False)
    t.add_column(no_wrap=True)
    t.add_column()
    for lab, val in items:
        t.add_row(f"[bold]{lab}[/bold]", val)
    return t


def _export_report(result_path: str) -> None:
    """Сохраняет весь выведенный в консоль текст в файл."""
    try:
        with open(result_path, "w", encoding="utf-8") as f:
            f.write(console.export_text())
        console.print(f"[dim]Результаты сохранены: [cyan]{result_path}[/cyan][/dim]")
    except Exception as e:
        console.print(f"[yellow]Не удалось сохранить файл: {e}[/yellow]")

File: crates/dpi-detector/src/views/summary.rs
//! Summary panel: one row per tested area with its counters.

use crate::i18n::{Messages, fmt_size, fmt_speed};
use dpi_core::probe::domains::DomainStats;
use dpi_core::probe::telegram::TelegramFullReport;

use crate::tui::widgets::{BOX_WIDTH, asc, panel_to_string, strip_ansi_len, wrap_ansi};

fn frac_sgr(ok: usize, total: usize) -> &'static str {
    if total == 0 || ok == total {
        "32"
    } else if ok == 0 {
        "31"
    } else {
        "33"
    }
}

pub struct SummaryData<'a> {
    pub run_dns: bool,
    pub dns: Option<&'a dpi_core::probe::dns_avail::DnsAvailStats>,
    pub domains: Option<&'a DomainStats>,
    pub tcp: Option<(usize, usize, usize, usize)>, // ok, blocked, mixed, total
    pub run_telegram: bool,
    pub telegram: Option<&'a TelegramFullReport>,
}

pub fn render_summary(data: &SummaryData, msg: &Messages) -> String {
    let mut items: Vec<(String, String)> = Vec::new();

    if data.run_dns {
        if let Some(d) = data.dns {
            let doh_c = frac_sgr(d.doh_ok, d.doh_total);
            let udp_c = frac_sgr(d.udp_ok, d.udp_total);
            let mut parts = vec![format!("\x1b[{}m{}/{} DoH\x1b[0m", doh_c, d.doh_ok, d.doh_total)];
            if d.dot_total > 0 {
                let dot_c = frac_sgr(d.dot_ok, d.dot_total);
                parts.push(format!("\x1b[{}m{}/{} DoT\x1b[0m", dot_c, d.dot_ok, d.dot_total));
            }
            parts.push(format!("\x1b[{}m{}/{} UDP\x1b[0m", udp_c, d.udp_ok, d.udp_total));
            items.push((msg.summary_dns_avail.to_string(), parts.join("  ")));
            if !d.hijacked_brands.is_empty() {
                if d.resolvers_total > 0 && d.hijacked_brands.len() >= d.resolvers_total {
                    items.push((msg.summary_resolver_hijack.to_string(), format!("\x1b[31m{}\x1b[0m", msg.summary_all)));
                } else {
                    items.push((
                        msg.summary_resolver_hijack.to_string(),
                        format!("\x1b[31m{}\x1b[0m", d.hijacked_brands.join(", ")),
                    ));
                }
            } else {
                items.push((msg.summary_resolver_hijack.to_string(), "\x1b[2m—\x1b[0m".to_string()));
            }
            if d.subst_total > 0 {
                if d.fakeip_sub > 0 {
                    items.push((
                        msg.summary_fakeip_resp.to_string(),
                        format!("\x1b[35m{}/{} UDP\x1b[0m", d.fakeip_sub, d.subst_total),
                    ));
                }
                let rest = d.subst_sub.saturating_sub(d.fakeip_sub);
                if rest > 0 {
                    let c = if rest == d.subst_total { "31" } else { "33" };
                    items.push((
                        msg.summary_ans_hijack.to_string(),
                        format!("\x1b[{}m{}/{} UDP\x1b[0m", c, rest, d.subst_total),
                    ));
                } else if d.fakeip_sub == 0 {
                    items.push((
                        msg.summary_ans_hijack.to_string(),
                        format!("\x1b[32m0/{} UDP\x1b[0m", d.subst_total),
                    ));
                }
            }
        } else {
            items.push((msg.summary_dns_avail.to_string(), "\x1b[2m—\x1b[0m".to_string()));
        }
    }

    if let Some(d) = data.domains {
        let stat = |label: &str, ok: usize| {
            let c = frac_sgr(ok, d.total);
            format!("\x1b[{}m{}/{} {}\x1b[0m", c, ok, d.total, label)
        };
        items.push((
            msg.summary_domains.to_string(),
            format!("{}  {}  {}", stat("HTTP", d.http_ok), stat("TLS1.2", d.t12_ok), stat("TLS1.3", d.t13_ok)),
        ));
    }

    if let Some((ok, blocked, mixed, total)) = data.tcp {
        let pct = ok.checked_mul(100).and_then(|v| v.checked_div(total)).unwrap_or(0);
        let mut value = format!("\x1b[32m√ {}/{} OK\x1b[0m", ok, total);
        if blocked > 0 {
            value += &format!("  \x1b[31m× {} {}\x1b[0m", blocked, msg.blocked_short);
        }
        if mixed > 0 {
            value += &format!("  \x1b[33m≈ {} {}\x1b[0m", mixed, msg.mixed_short);
        }
        value += &format!("  \x1b[2m({}% OK)\x1b[0m", pct);
        items.push(("TCP 16-20KB".to_string(), value));
    }

    if data.run_telegram {
        if let Some(t) = data.telegram {
            
            let tg_row = |label: &str, st: &dpi_core::probe::telegram::TransferStats, speed: f64, size: u64| {
                let (raw, sgr) = match st.status.as_str() {
                    "ok" => ("OK", "32"),
                    "stalled" => ("STALL", "33"),
                    "slow" => ("SLOW", "33"),
                    "blocked" => ("BLOCKED", "31"),
                    _ => ("ERROR", "31"),
                };
                let mut metrics = format!("{} {}, {}", msg.avg_label, fmt_speed(speed, msg.lang), fmt_size(size, msg.lang));
                if let Some(sec) = st.drop_at_sec {
                    metrics += &msg.stall_after.replace("{}", &sec.to_string());
                }
                (label.to_string(), format!("\x1b[{}m{:<16}\x1b[0m {}", sgr, raw, metrics))
            };
            let (l1, v1) = tg_row(msg.summary_tg_download, &t.download, t.download.avg_bps, t.download.bytes_total);
            let (l2, v2) = tg_row(msg.summary_tg_upload, &t.upload, t.upload.avg_bps, t.upload.bytes_total);
            items.push((l1, v1));
            items.push((l2, v2));
            let dc_c = if t.dc_reachable == t.dc_total {
                "32"
            } else if t.dc_reachable == 0 {
                "31"
            } else {
                "33"
            };
            items.push((
                msg.summary_tg_datacenters.to_string(),
                format!("\x1b[{}mOK {}/{}\x1b[0m", dc_c, t.dc_reachable, t.dc_total),
            ));
        }
    }

    if items.is_empty() {
        return String::new();
    }
    // Two columns: the label column never wraps, so pad it to its widest entry
    // and every value starts at the same offset. Widths are measured after ANSI
    // stripping, and per character, because CJK labels are two cells wide.
    let label_w = items.iter().map(|(label, _)| strip_ansi_len(label)).max().unwrap_or(0);
    // The value column: indent + label column + gap, and what is left of a row
    // once the borders and the single space around the content are taken out.
    let value_col = 2 + label_w + 2;
    let value_w = BOX_WIDTH.saturating_sub(3 + value_col);
    let mut lines = Vec::new();
    for (label, val) in items {
        let pad = label_w.saturating_sub(strip_ansi_len(&label));
        // `asc` widens glyphs in ASCII mode (`✓` becomes `[OK]`), so it has to run
        // before the width is measured; `panel_with` repeats it on the finished
        // line, which is idempotent.
        let value = asc(&val);
        for (i, chunk) in wrap_ansi(&value, value_w).into_iter().enumerate() {
            if i == 0 {
                lines.push(format!("  \x1b[1m{}{}\x1b[0m  {}", label, " ".repeat(pad), chunk));
            } else {
                lines.push(format!("{}{}", " ".repeat(value_col), chunk));
            }
        }
    }
    panel_to_string(msg.summary_title, &lines)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::tui::backend::strip_ansi;
    /// The summary is a two-column table (label, value): the label column is
    /// padded to its widest entry, so a short label cannot pull its value out of
    /// the column.
    #[test]
    fn summary_rows_align_their_values_into_two_columns() {
        use crate::i18n::{get_messages, Language};
        let msg = get_messages(Language::Ru);
        let out = render_summary(
            &SummaryData {
                run_dns: true,
                dns: None,
                domains: None,
                tcp: Some((104, 0, 0, 110)),
                run_telegram: false,
                telegram: None,
            },
            &msg,
        );
        let rows: Vec<Vec<char>> = out
            .lines()
            .map(strip_ansi)
            .filter(|l| l.starts_with('│'))
            .map(|l| l.chars().collect())
            .collect();
        assert_eq!(rows.len(), 2, "one row per item, no wrapping");
        // Border, one space, the two-space indent, the label column, the gap.
        let longest = msg.summary_dns_avail.chars().count();
        let short = "TCP 16-20KB".len();
        assert!(longest > short, "the two labels must differ in width: {longest} vs {short}");
        let value_col = 1 + 1 + 2 + longest + 2;
        for row in &rows {
            let line: String = row.iter().collect();
            assert_ne!(row[value_col], ' ', "the value starts in the column: {line:?}");
            let gap: String = row[value_col - 2..value_col].iter().collect();
            assert_eq!(gap, "  ", "the column gap is intact: {line:?}");
        }
        // The short label is padded up to the column, not left ragged.
        let short_end = 1 + 1 + 2 + short;
        let pad: String = rows[1][short_end..value_col].iter().collect();
        assert_eq!(pad, " ".repeat(value_col - short_end), "pad the short label to the column");
    }

    /// The panel is a fixed-width box: `panel_with` pads a row, it cannot reflow
    /// one, so a value wider than its column has to be wrapped by the summary
    /// itself - otherwise it pushes the right border off the line.
    #[test]
    fn summary_wraps_a_long_value_inside_the_box() {
        use dpi_core::probe::dns_avail::DnsAvailStats;
        use crate::i18n::{get_messages, Language};
        let msg = get_messages(Language::Ru);
        let brands: Vec<String> = [
            "Cloudflare IP 2", "Google", "Level 3", "Level 3 2", "MSK-IX", "OpenDNS", "XboxDNS",
            "НСДИ",
        ]
        .iter()
        .map(|b| b.to_string())
        .collect();
        let stats = DnsAvailStats {
            doh_ok: 33,
            doh_total: 37,
            dot_ok: 33,
            dot_total: 34,
            udp_ok: 44,
            udp_total: 50,
            hijacked_brands: brands.clone(),
            resolvers_total: 120,
            subst_sub: 43,
            subst_total: 44,
            fakeip_sub: 0,
            top_stub: None,
        };
        let out = render_summary(
            &SummaryData {
                run_dns: true,
                dns: Some(&stats),
                domains: None,
                tcp: None,
                run_telegram: false,
                telegram: None,
            },
            &msg,
        );
        let rows: Vec<String> =
            out.lines().map(strip_ansi).filter(|l| l.starts_with('│')).collect();
        // Every row is exactly the box width: nothing spills past the border.
        for row in &rows {
            assert_eq!(strip_ansi_len(row), BOX_WIDTH, "{row:?}");
        }
        // One row for DNS availability, two for the hijack list, one for the
        // answer substitution - the list is the only value too wide to fit.
        assert_eq!(rows.len(), 4, "{rows:#?}");
        let label_w = msg.summary_resolver_hijack.chars().count();
        let value_col = 1 + 1 + 2 + label_w + 2;
        let list: Vec<String> = rows[1..rows.len() - 1]
            .iter()
            .map(|row| {
                assert_eq!(row.chars().nth(value_col - 1), Some(' '), "gap before the value: {row:?}");
                assert_ne!(row.chars().nth(value_col), Some(' '), "a value starts in the column: {row:?}");
                let text: String = row.chars().skip(value_col).collect();
                text.trim_end_matches('│').trim_end().to_string()
            })
            .collect();
        assert!(rows[1].contains(msg.summary_resolver_hijack), "the label sits on the first line");
        assert!(!rows[2].contains(msg.summary_resolver_hijack), "the continuation repeats no label");
        assert_eq!(list.join(" "), brands.join(", "), "wrapping loses no entry");
    }
}
